src/inmanta/agent/forking_executor.py

```diff
@@ -685,179 +874,240 @@ def init_once(cls) -> None:
+
+    def _id_to_internal(self, ext_id: executor.ExecutorBlueprint) -> executor.ExecutorBlueprint:
+        return ext_id
+
+    def render_id(self, member: executor.ExecutorId) -> str:
+        return f"Executor for {member.agent_name}"
+
+    def my_name(self) -> str:
+        return "Executor Manager"
+
+    async def get_executor(
+        self,
+        agent_name: str,
+        agent_uri: str,
+        code: typing.Collection[executor.ResourceInstallSpec],
+    ) -> MPExecutor:
+        """
+        Retrieves an Executor for a given agent with the relevant handler code loaded in its venv.
+        If an Executor does not exist for the given configuration, a new one is created.
+
+        :param agent_name: The name of the agent for which an Executor is being retrieved or created.
+        :param agent_uri: The name of the host on which the agent is running.
+        :param code: Collection of ResourceInstallSpec defining the configuration for the Executor i.e.
+            which resource types it can act on and all necessary information to install the relevant
+            handler code in its venv.
+        :param venv_checkup_interval: The time interval after which the virtual environment must be touched. Only used for
+            testing. The default value is set to 60.0. It should not be used except for testing purposes. It can be
+            It can be overridden to speed up the tests
+        :return: An Executor instance
+        """
+        blueprint = executor.ExecutorBlueprint.from_specs(code)
+        executor_id = executor.ExecutorId(agent_name, agent_uri, blueprint)
+
+        my_executor = await self.get(executor_id)
+
+        if my_executor.failed_resource_results and not my_executor.failed_resources:
+            # If some code loading failed, resolve here
+            # reverse index
+            type_for_spec: dict[inmanta.loader.ModuleSource, list[ResourceType]] = collections.defaultdict(list)
+            for spec in code:
+                for source in spec.blueprint.sources:
+                    type_for_spec[source].append(spec.resource_type)
+            # resolve
+            for failed_resource_result in my_executor.failed_resource_results:
+                for rtype in type_for_spec.get(failed_resource_result.module_source, []):
+                    if rtype not in my_executor.failed_resources:
+                        my_executor.failed_resources[rtype] = failed_resource_result.exception
+
+        # FIXME: recovery. If loading failed, we currently never rebuild https://github.com/inmanta/inmanta-core/issues/7695
+        return my_executor
+
+    async def create_member(self, executor_id: executor.ExecutorId) -> MPExecutor:
+        process = await self.process_pool.get(executor_id.blueprint)
+        # TODO: we have a race here: the process can become empty between these two calls
+        result = await process.get(executor_id)
+        self.agent_map.get(executor_id.agent_name).add(result)
```

Comment:
type error I don't know how to resolve

```diff
src/inmanta/agent/forking_executor.py:1093: error: Item "None" of "set[MPExecutor] | None" has no attribute "add"  [union-attr]
src/inmanta/agent/forking_executor.py:1097: error: Item "None" of "set[MPExecutor] | None" has no attribute "discard"  [union-attr]

```

Reply:
The value of the dict could be saved in a variable and asserted to be not None.

Reply:
It is a bit silly default dict can't be typed more strictly, https://github.com/python/typeshed/tree/main/stdlib/collections#L399 because the factory argument is optional